### PR TITLE
feat(favorites): frontend revisions to pray and pay project

### PR DESCRIPTION
**Pull request by v-anne**
This is the PR for making frontend changes to the pray and pay user interface.

Mike had some great comments in #4507, so we'll start with them and see what's needed once the low-hanging fruit is done.

Additions include:
- Allowing users to request a document from the `recap_document.html` page.
- Revamping UI of `top_prayers.html` to provide better UX.
- Creating `user_prayers.html` page to allow users to see their open and fulfilled prayers, and also summary statistics about how many requests they've had fulfilled. In a future PR, this will be made similar to tags to allow people (e.g., freelance journalists) to share documents they've requested.
- documentation for how pray and pay project works.
- minor grammar changes to email notifications.

### CODE REVIEW

**Comment by v-anne** [comment]
@ERosendo this is the PR I'm going to build off of.

**Comment by mlissner** [comment]
Seeing some commit messages around the user-prayer page. Reminds me that it doesn't *have* to be available in the profile section, just like tags aren't. The user prayers page could just be linked from the mouseover for each emoji, for example.

**Comment by v-anne** [comment]
Yeah, tags are what I am aiming to emulate.

**Comment by v-anne** [comment]
image
I'm having some issues getting the three buttons to display inline with each other. Ideally, the prayer button would be to the right of the other two.

**Comment by v-anne** [comment]
image

Having same issue with the button on the modal. There's also a related issue where if I click the prayer button repeatedly, the count will increase despite having exhausted the quota.

**Comment by ERosendo** [comment]
@mlissner sure! 👍

**Comment by v-anne** [comment]
@ERosendo in addition to those two issues with the buttons, I've mostly finished a MVP of the `user_prayers.html` page. 

This is what it currently looks like. The status column is broken; I can't figure out why the backend isn't passing the information through to the template. I intend to add a button to allow users to delete requests and one for them to download available documents as well.

image

**Comment by v-anne** [comment]
@mlissner @ERosendo I've finished up my end of this. I've highlighted some concerns of mine regarding bits that might merit closer scrutiny.

**Comment by mlissner** [comment]
Can you post a few more screenshots for me, @v-anne?

**Comment by v-anne** [comment]
image
image

**Review by ERosendo — CHANGES_REQUESTED** [review]
@v-anne Here are my initial comments on the PR. I'll help you with the alignment issues in a bit.

File: cl/favorites/utils.py
Comment (mlissner) [reply cl/favorites/utils.py:265]: Seems like we need a rule or linter to this effect. It's really just too tempting to return a tuple like this when a dataclass is almost universally better. Hm.

**Comment by mlissner** [comment]
@v-anne, I noticed this PR is paused. Are you waiting for feedback or something else?

**Comment by v-anne** [comment]
Sorry @mlissner, I had stepped back because I thought @ERosendo was going to take a look at some of the issues I had raised with my code.

**Comment by mlissner** [comment]
Got it. We're moving to a sprint/scrum model, but I'll get this on his new list. :)

**Comment by v-anne** [comment]
Not related to these changes Mike, but I think there might be some issues with email. I'm not receiving alerts for every fulfilled request.

**Comment by mlissner** [comment]
> Not related to these changes Mike, but I think there might be some issues with email. I'm not receiving alerts for every fulfilled request.

I checked your account and we've sent you 41 emails about pay-and-pray so far. If you want to debug a particular one, emails is probably the better way to go.

**Comment by ERosendo** [comment]
Fixed alignment issue in `recap_document.html` and `redirect_to_pacer_modal.html` templates.


### recap_document.html:

image

----

image


----

image


### redirect_to_pacer_modal.html:

image

----

image

----

image

**Review by ERosendo — APPROVED** [review]

**Comment by ERosendo** [comment]
@v-anne Thanks for the hard work in this PR. 

@mlissner In addition to the alignment corrections addressed in comment https://github.com/freelawproject/courtlistener/pull/4580#issuecomment-2523385303, I've refined the help page for the Pray and Pay project to include accurate screenshots and updated the user list of prayers to utilize HTMX for the delete button.

Here's the updated help page:

Screen Recording 2024-12-16 at 2 45 09 PM


Here's a GIF showing the new user prayer list.

Screen Recording 2024-12-16 at 2 53 22 PM

**Comment by mlissner** [comment]
Seems OK to me, generally, thanks. Let's get one more set of eyes on this though, because there's a lot of code here. 

Thank you!

**Comment by v-anne** [comment]
It looks pretty good to my eyes. Local testing seemed alright as well.

I'll defer to another reviewer to decide whether it's ready to merge. I think this will be a great feature in the years ahead!

File: cl/favorites/templates/user_prayers.html
Comment (albertisfu) [inline cl/favorites/templates/user_prayers.html:52]: I found it a bit confusing that we call this `prayers`, but they are actually `RECAPDocument` instances. Should we rename these variables?

File: cl/favorites/templates/user_prayers.html
Comment (albertisfu) [inline cl/favorites/templates/user_prayers.html:58]: We can simplify this line to:

```suggestion
                   {{ docket|best_case_name|safe|v_wrapper }} ({{ docket.docket_number }})
```

File: cl/favorites/templates/user_prayers.html
Comment (albertisfu) [inline cl/favorites/templates/user_prayers.html:68]: This is displaying the `date_created` of the RECAPDocument. To display the `date_created` of the `Prayer`, we should use:

`{{ prayer.prayer_date_created|date:"M j, Y"  }}`

File: cl/favorites/templates/user_prayers.html
Comment (albertisfu) [inline cl/favorites/templates/user_prayers.html:77]: This line raises a `VariableDoesNotExist` error when loading the page with a `GRANTED` document because the `timezone` is not defined.

If we are going to display `date_upload` in UTC, we could use the custom filter: `datetime_in_utc`:

`{{ prayer.date_upload|datetime_in_utc }}`

This issue is also present in `prayer_email.html`

which could be the reason some emails are being missed, as seen in sentry: [COURTLISTENER-8FC](https://freelawproject.sentry.io/issues/6026700048/)

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:132]: Since `best_case_name` can also use `case_name_full` and `case_name_short`. We should add them here to avoid additional queries.

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:179]: Similarly to `get_top_prayers` we should include `case_name_full` and `case_name_short`.

Additionally, some other fields are being used in the template:
```
date_upload
filepath_local
filepath_ia
```

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:262]: I'm concerned about the performance of this `total_cost` computation if the user has thousands of prayers.

Would it be possible to use a similar approach, relying on a database query like the one you used in `get_lifetime_prayer_stats` to compute the cost?

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:308]: I think it is possible to simplify this annotation a bit by relying on `Least`:

```suggestion
            price=Case(
                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
                When(
                    recap_document__page_count__gt=0,
                    then=Least(
                        Value(3.0),
                        F("recap_document__page_count") * Value(0.10)
                    ),
                ),
                default=Value(0.0),
            )
```

I would also suggest adding a test for this method since it is a bit complex. This will help guarantee its functionality and ease future maintainability.

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:321]: We are caching the `data` here but it seems we are not using the previously cached values if they exist to avoid running the query again

File: cl/favorites/views.py
Comment (albertisfu) [inline cl/favorites/views.py:290]: Similar to a previous comment, I got a bit confused since this method returns `RECAPDocuments`. Should we rename it to something like `rd_prayers`?

File: cl/opinion_page/views.py
Comment (albertisfu) [inline cl/opinion_page/views.py:746]: It seems this context variable is not being used in any template. It's also present in `view_docket`.

**Review by albertisfu — CHANGES_REQUESTED** [review]
This looks great. I really like how this feature is taking shape!

I've left some comments and suggestions above.

Additionally, I think it would be a good idea to add pagination to `user_prayers_view`, considering that some users may have thousands of prayers, which can impact performance. Or not sure if you're already considering this but it's not within the scope of this PR.

File: cl/favorites/templates/user_prayers.html
Comment (ERosendo) [reply cl/favorites/templates/user_prayers.html:58]: You're right! Thanks for catching this.

File: cl/favorites/templates/user_prayers.html
Comment (ERosendo) [reply cl/favorites/templates/user_prayers.html:52]: Given that we're iterating over documents with the highest prayer count, I think we can name the variable `rd_with_prayers`. What are your thoughts?

File: cl/favorites/templates/user_prayers.html
Comment (ERosendo) [reply cl/favorites/templates/user_prayers.html:68]: You're correct!

File: cl/favorites/templates/user_prayers.html
Comment (ERosendo) [reply cl/favorites/templates/user_prayers.html:77]: Great catch. Thanks!

File: cl/favorites/templates/user_prayers.html
Comment (albertisfu) [reply cl/favorites/templates/user_prayers.html:52]: yeah I agree!

File: cl/favorites/utils.py
Comment (ERosendo) [reply cl/favorites/utils.py:132]: alright!

File: cl/favorites/views.py
Comment (ERosendo) [reply cl/favorites/views.py:290]: I apologize, I unintentionally removed this in commit [5fcb609](https://github.com/freelawproject/courtlistener/pull/4580/commits/5fcb609666bdb8e215391f60850c19c1ca1d91f4)

File: cl/favorites/utils.py
Comment (ERosendo) [reply cl/favorites/utils.py:321]: I apologize, I unintentionally removed this in commit [5fcb609](https://github.com/freelawproject/courtlistener/pull/4580/commits/5fcb609666bdb8e215391f60850c19c1ca1d91f4)

File: cl/favorites/views.py
Comment (ERosendo) [reply cl/favorites/views.py:290]: I agree. I'll rename this variable

File: cl/opinion_page/views.py
Comment (ERosendo) [reply cl/opinion_page/views.py]: You're correct! Thanks for bringing this to my attention.

File: cl/favorites/utils.py
Comment (ERosendo) [reply cl/favorites/utils.py:269]: > Would it be possible to use a similar approach, relying on a database query like the one you used in get_lifetime_prayer_stats to compute the cost?

Yes, that's possible. I'll update this helper function accordingly.

**Comment by ERosendo** [comment]
@albertisfu Thanks for your feedback. I've implemented all your suggestions.

File: cl/favorites/templates/prayer_email.html
Comment (albertisfu) [inline cl/favorites/templates/prayer_email.html:48]: In this line: `{{ docket_entry.datetime_filed|datetime_in_utc|date:"M j, Y" }}`, an empty value is displayed.

I believe the issue is that the `datetime_in_utc` filter handles the date format internally, so applying an additional format causes it to fail. If we want to display a different format here, we might need to use a different approach, such as modifying `datetime_in_utc` to accept an optional custom format.

Same for the .txt version

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:285]: I think we can avoid the duplicate code in this method and in `get_lifetime_prayer_stats` by encapsulating the shared logic into a separate method, something like:

```
async def compute_cost(queryset):
    return await (
        queryset.values("recap_document")
        .distinct()
        .annotate(
            price=Case(
                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
                When(
                    recap_document__page_count__gt=0,
                    then=Least(
                        Value(3.0),
                        F("recap_document__page_count") * Value(0.10),
                    ),
                ),
                default=Value(0.0),
            )
        )
        .aaggregate(Sum("price", default=0.0))
    )
```

And then just using:
`cost =  await compute_cost(filtered_list)`

OR:

`total_cost = await compute_cost(prayer_by_status.select_related("recap_document"))
`

File: cl/favorites/utils.py
Comment (albertisfu) [inline cl/favorites/utils.py:348]: Can we use the same key names here as in `PrayerStats`?

I believe if we do something like:

```suggestion
        "prayer_count": prayer_count,
        "distinct_count": distinct_prayers,
        "total_cost": f"{total_cost:,.2f}",
```

Then in both returns we can do:
`return PrayerStats(**data)`

**Review by albertisfu — COMMENTED** [review]
Thanks @ERosendo this looks great.

Just left a couple of additional comments and suggestions.

File: cl/favorites/templates/prayer_email.html
Comment (ERosendo) [reply cl/favorites/templates/prayer_email.html:48]: we can pass `timezone` as a template variable and use this custom format, right?

File: cl/favorites/utils.py
Comment (ERosendo) [reply cl/favorites/utils.py:285]: You're correct! I'll create this new helper function.

File: cl/favorites/utils.py
Comment (ERosendo) [reply cl/favorites/utils.py:348]: You're correct! Thanks for bringing this to my attention.

File: cl/favorites/templates/prayer_email.html
Comment (albertisfu) [reply cl/favorites/templates/prayer_email.html:48]: yeah, that's also a valid approach!

**Comment by ERosendo** [comment]
@albertisfu Ready for another review 😃

**Review by albertisfu — APPROVED** [review]
Thanks! this now looks good to me :shipit:

**Comment by mlissner** [comment]
Right on. Into the maw it goes!! Thank you all for the teamwork polishing and refining this one. I'm looking forward to giving it a try.

**Pull request merged**
